Added Polymorphism in Golang (#3252)
girlscript/winter-of-contributing: Golang/9 Object Oriented Programming/9.3 Polymorphism/DOC.md

# <div align="center">🧲 Polymorphism in GoLang</div>
<p align="center"><img src="https://miro.medium.com/max/1200/1*BruAhzl61NYiVz83ZvEcEg.jpeg" alt="polymorphism" height=400x width=850px/></p>

> In Object-Oriented Programming, an object can behave like another object. This property is called polymorphism. This post will cover how we achieve polymorphism in GoLang.

Reviewer: ```suggestion
In Object-Oriented Programming, an object can behave like another object. This property is called **Polymorphism**. This post will cover how we achieve polymorphism in GoLang.

```

> In Object-Oriented Programming, an object can behave like another object. This property is called **Polymorphism**. This post will cover how we achieve polymorphism in GoLang.

Reviewer: Why this is not rendering perfectly?

```suggestion
</br>
In Object-Oriented Programming, an object can behave like another object. This property is called **Polymorphism**. This post will cover how we achieve polymorphism in GoLang.

```

Author: Done @amino19

> ## 🤔 What is polymorphism?

Reviewer: ```suggestion
## 🤔 What is Polymorphism?

```

> Polymorphism is a property that is available to many Object Oriented-languages. Go despite not being an OO-language achieves polymorphism through [interfaces](https://golangdocs.com/interfaces-in-golang).

Reviewer: ```suggestion
**Polymorphism** is a property that is available to many Object Oriented-languages. Go despite not being an OO-language achieves polymorphism through [interfaces](https://golangdocs.com/interfaces-in-golang).

```

Here is an example of polymorphism in action.

```go
package main
 
import "fmt"
 
// declare interface
type Dog interface {
    Bark()
}
 
// declare struct
type Dalmatian struct {
    DogType string
}
 
// implement the interface
func (d Dalmatian) Bark() {
    fmt.Println("Dalmatian barking!!")
}
 
func MakeDogBark(d Dog) {
    d.Bark()
}
 
func main() {
    d := Dalmatian{"Jack"}
    MakeDogBark(d)                    // Dalmatian barking!!
}
```
In the code above, the struct Dalmatian implements the Dog interface. Thus the struct becomes the type Dog and so that it can be passed in that function.
Now, we can simply add any type and implement that interface and the type will behave as that interface. That is polymorphism. An object taking many different forms.

> ## ✨Polymorphism using interfaces

Reviewer: ```suggestion
## ✨ Polymorphism using interfaces

```

    "fmt"
)

type Income interface {
    calculate() int
    source() string
}

type FixedBilling struct {
    projectName string
    biddedAmount int
}

type TimeAndMaterial struct {
    projectName string
    noOfHours  int
    hourlyRate int
}

func (fb FixedBilling) calculate() int {
    return fb.biddedAmount
}

func (fb FixedBilling) source() string {
    return fb.projectName
}

func (tm TimeAndMaterial) calculate() int {
    return tm.noOfHours * tm.hourlyRate
}

func (tm TimeAndMaterial) source() string {
    return tm.projectName
}

func calculateNetIncome(ic []Income) {
    var netincome int = 0
    for _, income := range ic {
        fmt.Printf("Income From %s = $%d\n", income.source(), income.calculate())
        netincome += income.calculate()
    }
    fmt.Printf("Net income of organization = $%d", netincome)
}

func main() {
    project1 := FixedBilling{projectName: "Project 1", biddedAmount: 5000}
    project2 := FixedBilling{projectName: "Project 2", biddedAmount: 10000}
    project3 := TimeAndMaterial{projectName: "Project 3", noOfHours: 160, hourlyRate: 25}
    incomeStreams := []Income{project1, project2, project3}
    calculateNetIncome(incomeStreams)
}
```

> This program will output

Reviewer: ```suggestion
This program will output:-

```

```go
type Advertisement struct {
    adName     string
    CPC        int
    noOfClicks int
}

func (a Advertisement) calculate() int {
    return a.CPC * a.noOfClicks
}

func (a Advertisement) source() string {
    return a.adName
}
```

The `Advertisement` type has three fields `adName`, `CPC` (cost per click) and `noOfClicks` (number of clicks). The total income from ads is the product of `CPC` and `noOfClicks`.

Let's modify the `main` function a little to include this new income stream.

```go
func main() {
    project1 := FixedBilling{projectName: "Project 1", biddedAmount: 5000}
    project2 := FixedBilling{projectName: "Project 2", biddedAmount: 10000}
    project3 := TimeAndMaterial{projectName: "Project 3", noOfHours: 160, hourlyRate: 25}
    bannerAd := Advertisement{adName: "Banner Ad", CPC: 2, noOfClicks: 500}
    popupAd := Advertisement{adName: "Popup Ad", CPC: 5, noOfClicks: 750}
    incomeStreams := []Income{project1, project2, project3, bannerAd, popupAd}
    calculateNetIncome(incomeStreams)
}
```

We have created two ads namely `bannerAd` and `popupAd`. The `incomeStreams` slice includes the two ads we just created.

> Here is the full program after adding Advertisement.

Reviewer: ```suggestion
Here is the full program after adding Advertisement:

```

Reviewer: @eshan1925: Pl do this change! :eyes:

    project2 := FixedBilling{projectName: "Project 2", biddedAmount: 10000}
    project3 := TimeAndMaterial{projectName: "Project 3", noOfHours: 160, hourlyRate: 25}
    bannerAd := Advertisement{adName: "Banner Ad", CPC: 2, noOfClicks: 500}
    popupAd := Advertisement{adName: "Popup Ad", CPC: 5, noOfClicks: 750}
    incomeStreams := []Income{project1, project2, project3, bannerAd, popupAd}
    calculateNetIncome(incomeStreams)
}
```

The above program will output,

```
Income From Project 1 = $5000
Income From Project 2 = $10000
Income From Project 3 = $4000
Income From Banner Ad = $1000
Income From Popup Ad = $3750
Net income of organization = $23750
```

You would have noticed that we did not make any changes to the `calculateNetIncome` function though we added a new income stream. It just worked because of polymorphism. Since the new `Advertisement` type also implemented the `Income` interface, we were able to add it to the `incomeStreams` slice. The `calculateNetIncome` function also worked without any changes as it was able to call the `calculate()` and `source()` methods of the `Advertisement` type.

> 

Reviewer: Reduce the number of lines here! :eye:

Reviewer: @eshan1925: Pl do this change!

## 🚨 Uses of polymorphism

Polymorphism is used to reduce code in general. There will be less coupling if polymorphism is used. A single function can be used to do the same thing on multiple different objects. This is where polymorphism is heavily used. It is one of the most important concepts in OO-Programming. Go not being a strict OO-language achieves polymorphism in an elegant way.

## Few more Examples of Polymorphism

> ### Example-1:

Reviewer: ```suggestion
### Example-1:-

```

Reviewer: @eshan1925: Pl do this change! :eyes:

```go

> // Golang program to illustrate the

Reviewer: ```suggestion
/* Golang program to illustrate the

```

> // concept of interfaces

Reviewer: ```suggestion
   concept of interfaces */

```

Author: Ok anything else? Or this is all i need to do😁

package main

import (
	"fmt"
)

// defining an interface
type Figure interface{

	Area() float64
}

// declaring a struct
type Rectangle struct{
	
	// declaring struct variables
	length float64
	width float64
}

// declaring a struct
type Square struct{
	
	// declaring struct variable
	side float64
}

// function to calculate
// area of a rectangle
func (rect Rectangle) Area() float64{

	// Area of rectangle = l * b
	area := rect.length * rect.width
	return area
}

// function to calculate
// area of a square
func (sq Square) Area() float64{
	
	// Area of square = a * a
	area := sq.side * sq.side
	return area
}

// main function
func main() {
	
	// declaring a rectangle instance
	rectangle := Rectangle{
	
		length: 10.5,
		width: 12.25,
	}
	
	// declaring a square instance
	square := Square{
	
		side: 15.0,
	}
	
	// printing the calculated result
	fmt.Printf("Area of rectangle: %.3f unit sq.\n", rectangle.Area())
	fmt.Printf("Area of square: %.3f unit sq.\n", square.Area())
}
```

> #### Output:

Reviewer: ```suggestion
#### Output:-

```

Reviewer: @eshan1925: Pl do this change! :eyes:

```
Area of rectangle: 128.625 unit sq.
Area of square: 225.000 unit sq.
```

Objects of different types are treated in a consistent way, as long as they stick to a single interface, which is the essence of polymorphism. Variable declared in an interface are of interface type. They can take whichever value which implements the interface which helps interfaces to achieve polymorphism in the Golang. The following example explains the concept of polymorphism:

> ### Example-2:

Reviewer: ```suggestion
### Example-2:-

```

Reviewer: @eshan1925: Pl do this change! :eyes:

	return read_time
}

// function to calculate reading
// time for magazine
func (magazine Magazine) reading_time() int {
	
	// taking average speed
	// of 5 mins per page
	read_time := 5 * magazine.page_count
	return read_time
}

// function to calculate reading time
func calcReadingTime(ReadingTime []Reading) int {

		totalTime := 0
	
	// looping through elements
	// of the Reading array
		for _, t := range ReadingTime {
		
		// run time polymorphism, call to
		// method depends on object being
		// referred at run time
			totalTime += t.reading_time()
		}
	
		return totalTime
}

// main function
func main() {
	
	// declaring a book instance
	book1 := Book{
		name: "Goosebumps",
		page_count: 150,
	}
	
	// declaring a newspaper instance
	newspaper1 := Newspaper{
		name: "TOI",
		page_count: 12,
	}
	
	// declaring a magazine instance
	magazine1 := Magazine{
		name: "Forbes",
		page_count: 40,
	}
	
	// array of type Reading interface
	ReadingTime := []Reading{book1, newspaper1, magazine1}
	
	// total reading time calculated
		totalTime := calcReadingTime(ReadingTime)
	
	// Printing total time for reading
		fmt.Printf("Total Time is %d minutes.\n", totalTime)
}
```

> #### Output:

Reviewer: ```suggestion
#### Output:-

```

Reviewer: @eshan1925: Pl do this change! :eyes:

```
Total Time is 2060 minutes.
```

> So thats all that you need to knwo about Polymorphism in Golang if you wish to know more you can definetly visit the official documentation of <a href="https://golang.org/doc/">Golang</a>.

Reviewer: ```suggestion
So that's all that you need to know about Polymorphism in Golang if you wish to know more you can definitely visit the official documentation of <a href="https://golang.org/doc/">Golang</a>.

```

> ### 🛠 References-: 

Reviewer: ```suggestion
### 🛠 References:- 

```

> - The above documentation was prepared by reffering to the <a href="https://golang.org/doc/">official documentation</a> of Golang for authencity, <a href="https://www.geeksforgeeks.org/">Geeks for geeks</a> and some <a href="https://golangbot.com/polymorphism/">blogs</a>for explainging the concepts in detail.

Reviewer: ```suggestion
- The above documentation was prepared by referring to the [official documentation](https://golang.org/doc/) of Golang for authenticity, [Geeks for geeks](https://www.geeksforgeeks.org/) and some [blogs](https://golangbot.com/polymorphism/) for explaining the concepts in detail.

```

Reviewer: @eshan1925: Removed the link, ig. Why?

> - The images that are used are also used form the following links-:

Reviewer: ```suggestion
- The images that are used are also used from the following links:-

```

Reviewer: @eshan1925: Still not resolved!

 <table align="center">
  <tr>
   <td>
    <a href="">Link1</a></td>
   <td>
    <a href="">Link2</a></td>
    </tr>
  </table>

> ## THANKS FOR READING! 😎✨<br>

Reviewer: ```suggestion
## THANKS FOR READING! ✨<br>

```
